src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java

```diff
@@ -17,33 +17,252 @@
+      String taskId) {
     return metric
         .setState(state)
-        .setTaskId(taskId)
-        .setExpectedStateCounts(stateCounts)
+        .setTaskId(UUID.fromString(taskId))
         // TODO use metric timestamps instead of `now`` ?
         .setStateEvaluationTimestamp(Instant.now());
   }
 
+  public static SalusEnrichedMetric generateEnrichedMetric(
+      SalusEnrichedMetric metric, String taskId) {
+    EsperTaskData data = taskDataMap.get(taskId);
+    Map<String, Double> customMetricData = new HashMap<>();
+
+    //prep custom metrics
+    List<MetricExpressionBase> customMetrics =
+        data.getEventEngineTask().getTaskParameters().getCustomMetrics();
+    if (customMetrics != null) {
+      StateEvaluator.MetricEvaluator evaluator = new StateEvaluator.MetricEvaluator(
+          metric.getMetrics());
+      for (MetricExpressionBase customMetric : customMetrics) {
+        Map.Entry<String, Double> entry = evaluator.eval(customMetric);
+        customMetricData.put(entry.getKey(), entry.getValue());
+      }
+    }
+    // sort state expressions
+    Map<TaskState, StateExpression> stateExpressions = new HashMap<>();
+    for (StateExpression stateExpression : data.getEventEngineTask().getTaskParameters()
+        .getStateExpressions()) {
+      stateExpressions.put(stateExpression.getState(), stateExpression);
+    }
+
+    // Check each type of stateExpression and return the first one that is true
+    for (String state : List.of("CRITICAL", "WARNING", "OK")) {
+      StateExpression current = stateExpressions.get(TaskState.valueOf(state));
+      if (current != null && getStateExpressionValue(current, metric, customMetricData)) {
+        return initMetric(metric, state, taskId );
+      }
+    }
+    // None were true so return ok
+    return initMetric(metric, "OK", taskId);
+  }
+
+  private static boolean getStateExpressionValue(StateExpression current,
+      SalusEnrichedMetric metric, Map<String, Double>customMetricData) {
+    if (current.getExpression() instanceof LogicalExpression) {
+      return getLogicalValue((LogicalExpression) current.getExpression(), metric, customMetricData);
+    } else {
+      return getComparisonValue((ComparisonExpression) current.getExpression(), metric,
+          customMetricData);
+    }
+  }
+
+  private static boolean getLogicalValue(LogicalExpression expression,
+      SalusEnrichedMetric metric, Map<String, Double>customMetricData) {
+    boolean retval;
+    // With Operator.AND return as soon as false is found
+    if (expression.getOperator() == Operator.AND) {
+      for (Expression subExpression : expression.getExpressions()) {
+        if (subExpression instanceof LogicalExpression) {
+          retval = getLogicalValue((LogicalExpression) subExpression, metric, customMetricData);
+
+        } else {
+          retval = getComparisonValue((ComparisonExpression) subExpression,
+              metric, customMetricData);
+        }
```

Comment:
```suggestion
        retval = getStateExpressionValue(subExpression, metric, customMetricData);
```

Can something like this not be used instead?